elks/arch/i86/drivers/char/console-direct.c

```diff
@@ -183,54 +246,114 @@ struct tty_ops dircon_ops = {
     Console_conout
 };
 
-void INITPROC console_init(void)
+
+/* Check to see if this CRTC is present */
+static char probe_crtc(unsigned short crtc_base)
 {
-    Console *C;
     int i;
-    unsigned int PageSizeW;
-
-    MaxCol = (Width = peekb(0x4a, 0x40)) - 1;  /* BIOS data segment */
-
-    /* Trust this. Cga does not support peeking at 0x40:0x84. */
-    MaxRow = (Height = 25) - 1;
-    CCBase = peekw(0x63, 0x40);
-    PageSizeW = ((unsigned int)peekw(0x4C, 0x40) >> 1);
-
-    VideoSeg = 0xb800;
-    if (peekb(0x49, 0x40) == 7) {
-        VideoSeg = 0xB000;
-        NumConsoles = 1;
-        isMDA = 1;
-    } else {
-        isCGA = peekw(0xA8+2, 0x40) == 0;
+    unsigned char test = 0x55;
+    /* We'll try writing a value to cursor address LSB reg (0x0F), then reading it back */
+    outb(0x0F, crtc_base + CRTC_INDX);
+    unsigned char original = inb(crtc_base + CRTC_DATA);
+    if (original == test) test += 0x10;
+    outb(0x0F, crtc_base + CRTC_INDX);
+    outb(test, crtc_base + CRTC_DATA);
+    /* Now wait a bit */
+    for (i = 0; i < 100; ++i) {} // TODO: Verify this doesn't get optimized out
+    outb(0x0F, crtc_base + CRTC_INDX);
+    unsigned char value = inb(crtc_base + CRTC_DATA);
+    if (value != test) return 0;
+    outb(0x0F, crtc_base + CRTC_INDX);
+    outb(original, crtc_base + CRTC_DATA);
+    return 1;
+}
+
+static int init_output(enum OutputType t)
+{
+    int i;
+    struct hw_params *p = &params[t];
+    /* Set 80x25 mode, video off */
+    outb(0x01, p->crtc_base + CRTC_MODE);
+    /* Program CRTC regs */
+    for (i = 0; i < p->n_init_bytes; ++i) {
+        outb(i, p->crtc_base + CRTC_INDX);
+        outb(p->init_bytes[i], p->crtc_base + CRTC_DATA);
     }
 
-    C = Con;
-    Visible = C;
+    /* Check & clear vram */
+    for (i = 0; i < p->vseg_bytes; i += 2)
+        pokew(i, p->vseg_base, 0x5555);
+    for (i = 0; i < p->vseg_bytes; i += 2)
+        if (peekw(i, p->vseg_base) != 0x5555) return 1;
+    for (i = 0; i < p->vseg_bytes; i += 2)
+        pokew(i, p->vseg_base, 0x7 << 8 | ' ');
+
+    /* Enable video */
+    outb(0x09, p->crtc_base + CRTC_MODE);
+    return 0;
+}
 
-    for (i = 0; i < NumConsoles; i++) {
-        C->cx = C->cy = 0;
-        if (!i) {
-            C->cx = peekb(0x50, 0x40);
-            C->cy = peekb(0x51, 0x40);
-        }
-        C->fsm = std_char;
-        C->basepage = i * PageSizeW;
-        C->vseg = VideoSeg + (C->basepage >> 3);
-        C->attr = A_DEFAULT;
+static const char *type_string(enum OutputType t)
+{
+    switch (t) {
+    case OT_MDA: return "MDA";
+    case OT_CGA: return "CGA";
+    case OT_EGA: return "EGA";
+    case OT_VGA: return "VGA";
```

Comment:
No need for code here - the strings can be in a static char * array indexed by OT_xxx, used as a %s in printk rather than calling a function to get a string.

We don't differentiate in ELKS between EGA and VGA so perhaps this also should be gated. Although I very much like this new enhancement, I have been trying to reduce the base requirements for ELKS, not increase them. It has been a long road...

Reply:
Very good call! I feel a bit silly now, having written that :D